src/index.ts

```diff
@@ -47,52 +47,42 @@ interface ConnectSocketOptions {
 
 // Connect an http socket to another tcp server.
 // Based on tunneling proxy code from https://nodejs.org/api/http.html
-const connectSocket = ({
+const connectSocket = async ({
 	cltSocket,
 	hostname,
 	port,
 	head,
 	connect,
 	req,
-}: ConnectSocketOptions) =>
-	connect(
-		port,
-		hostname,
-		cltSocket,
-		req,
-	)
-		.then(srvSocket => {
-			cltSocket.write(
-				`HTTP/1.0 200 Connection Established\r\nProxy-agent: balena-io/node-tunnel (v${version})\r\n\r\n`,
-			);
-			srvSocket.write(head);
-			srvSocket.pipe(
-				cltSocket,
-				{ end: false },
-			);
-			cltSocket.pipe(
-				srvSocket,
-				{ end: false },
-			);
-
-			return Promise.fromCallback(cb => {
-				cltSocket.on('error', cb);
-				srvSocket.on('error', cb);
-				cltSocket.on('end', cb);
-				srvSocket.on('end', cb);
-			}).finally(() => {
-				srvSocket.destroy();
-				cltSocket.destroy();
-			});
-		})
-		.tapCatch(() => {
-			if (cltSocket.writable) {
-				cltSocket.end('HTTP/1.0 500 Internal Server Error\r\n');
-			}
-			if (!cltSocket.destroyed) {
-				cltSocket.destroy();
-			}
+}: ConnectSocketOptions) => {
+	try {
+		const srvSocket = await connect(port, hostname, cltSocket, req);
+		cltSocket.write(
+			`HTTP/1.0 200 Connection Established\r\nProxy-agent: balena-io/node-tunnel (v${version})\r\n\r\n`,
+		);
+		srvSocket.write(head);
+		srvSocket.pipe(cltSocket, { end: false });
+		cltSocket.pipe(srvSocket, { end: false });
+
+		return Bluebird.fromCallback(cb => {
+			cltSocket.on('error', cb);
+			srvSocket.on('error', cb);
+			cltSocket.on('end', cb);
+			srvSocket.on('end', cb);
+		}).finally(() => {
```

Comment:
This can be an awaited `try/finally`

Reply:
I want to return before the promise resolves though, no?

Reply:
You'll be returning a pending promise either way, this function technically returns the moment you hit the first await in `const srvSocket = await connect(port, hostname, cltSocket, req);`